### ISSUE TITLE
Training with new dataset

### ISSUE THREAD

**Issue by MingZJU** [primepake/wav2lip_288x288#10]
Hi, Happy Christmas!

I'd like to train the model with hq images (320X320) and need some help.

There is a loss error in line 133 of "color_syncnet_train.py".  The loss is out of (0, 1).

It seems work when I change loss function 
from
logloss = nn.BCELoss()
to
logloss = nn.BCEWithLogitsLoss()

Is it OK? Do I have to make other changes?

The training lasts more than 4 days now, loss around 0.54 after 940, 000 steps.

Thanks in advance.

**Comment by primepake** [comment]
I've used logloss = nn.BCEWithLogitsLoss() to solve this problem. But then I realized it tended to vanishing problem again. So I decided to apply Wasserstein distance loss

**Comment by MingZJU** [comment]
Many thanks, I'll try this loss.

**Comment by MingZJU** [comment]
I recently trained this model, but the discriminator's eval  loss is only 0.30, can not go down to ~0.25. 
Is it reasonable?
Thanks a lot.

**Comment by primepake** [comment]
let me know your batch_size, learning_rate, and logging when you're training

**Comment by primepake** [comment]
you can give me some of your sample?

**Comment by MingZJU** [comment]
Sorry for late reply.
Parameters: batch size 128, learning rate 5e-5, AdamW.

One sample is:

https://user-images.githubusercontent.com/53510802/160985858-a20d4f39-b26b-424b-a861-14b9ed2a2bfa.mp4

**Comment by primepake** [comment]
have you used Wasserstein distance or BCE loss? 
did you scale up the audio block in wav2lip model?

**Comment by MingZJU** [comment]
I used both ReLU and BCELoss. I also tried Wasserstein distance, but even worse.

what is your meaning of "scale up audio block"?

**Comment by primepake** [comment]
why don't you use leaky instead of relu?
because relu lead to vanishing problem
scale up audio block, I mean you need to add more hidden layers

**Comment by primepake** [comment]
ok, if you need some help, you can ask this thread

**Comment by MingZJU** [comment]
Can you tell me how to compute Wasserstein distance in syncnet?

def cosine_loss(a, v, y):
    d = nn.functional.cosine_similarity(a, v)
    loss = torch.mean(torch.squeeze(y)) - torch.mean(d)

There are other methods like torch.mean(y*d) or use scipy.stats.wasserstein_distance, but seems not right.

**Comment by primepake** [comment]
you need to understand theory and implement from scratch

**Comment by primepake** [comment]
hi, what is your infrastructure? 10 GPUS? telsa v100?

**Comment by MingZJU** [comment]
now I use 4 * V100

**Comment by primepake** [comment]
with batch size 128, How much memory does it take up?

**Comment by MingZJU** [comment]
With batch size 64, it takes less than 13G.
you cannot set 128 unless using Automatic Mixed Precision or your V100 with 32GB memory.

My V100 is with 16GB memery.

**Comment by primepake** [comment]
with color_syncnet_train.py, I trained on batch_size 128, lr=1e5, it has converged to 0.23
you need to use [syncnet](https://github.com/joonson/syncnet_python) to evaluate your dataset. the score need to be in range [-5,5]

**Comment by primepake** [comment]
I'm checking my wasserstein distance loss

**Comment by MingZJU** [comment]
thanks.
I do use the syncnet to correct my datasets.
The loss is only 0.28, I guess my wasserstein loss is still not right.

**Comment by primepake** [comment]
can you show me your Wasserstein loss?

**Comment by MingZJU** [comment]
I followed this [blog](https://machinelearningmastery.com/how-to-code-a-wasserstein-generative-adversarial-network-wgan-from-scratch/).
the loss is very simple: torch.mean(torch.mul(y_true, y_pred)), not right.
I may modify it from scratch in spare time.

**Comment by primepake** [comment]
good job bro :) I can see your result is very good

**Comment by NikitaKononov** [comment]
> Many thanks, I will scale up audio block.

Hello, can you please suggest how to scale audio block and does it make sence?

**Comment by NikitaKononov** [comment]
> why don't you use leaky instead of relu?

Hello)
I can see PReLU activation func in Conv2d and Conv2dTranspose
LeakyReLU is in nonorm_Conv2d
Original implementation uses ReLU
Where should I place LeakyReLU? in all blocks?

**Comment by GodofOrange** [comment]
> 



> I followed this [blog](https://machinelearningmastery.com/how-to-code-a-wasserstein-generative-adversarial-network-wgan-from-scratch/). the loss is very simple: torch.mean(torch.mul(y_true, y_pred)), not right. I may modify it from scratch in spare time.
> 
> This example genereated using bceloss and relu.
> 
>  result_voice.mp4

I would like to buy this model, how much is it ? connect me with 1243137612@qq.com @MingZJU

**Comment by Curisan** [comment]
> I followed this [blog](https://machinelearningmastery.com/how-to-code-a-wasserstein-generative-adversarial-network-wgan-from-scratch/). the loss is very simple: torch.mean(torch.mul(y_true, y_pred)), not right. I may modify it from scratch in spare time.
> 
> This example genereated using bceloss and relu.
> 
>  result_voice.mp4

Do you train the model using AVSpeech dataset?

**Comment by Curisan** [comment]
> I followed this [blog](https://machinelearningmastery.com/how-to-code-a-wasserstein-generative-adversarial-network-wgan-from-scratch/). the loss is very simple: torch.mean(torch.mul(y_true, y_pred)), not right. I may modify it from scratch in spare time.
> 
> This example genereated using bceloss and relu.
> 
>  result_voice.mp4

Do you train the model using AVSpeech dataset? @MingZJU

**Comment by MingZJU** [comment]
you can follow Primepake's code and obtain your results.I downloaded avspeech but have no time to train a model.-------- 原始邮件 --------发件人： Curisan ***@***.***>日期： 2023年1月29日周日 14:58收件人： primepake/wav2lip_288x288 ***@***.***>抄送： MingZJU ***@***.***>, Mention ***@***.***>主    题： Re: [primepake/wav2lip_288x288] Training with new dataset (Issue #10)

I followed this blog. the loss is very simple: torch.mean(torch.mul(y_true, y_pred)), not right. I may modify it from scratch in spare time.
This example genereated using bceloss and relu.
result_voice.mp4

Do you train the model using AVSpeech dataset? @MingZJU

**Comment by Curisan** [comment]
Thanks. I use AVSpeech dataset. And I did the following:
1. choose the 25fps video
2. using syncnet_python to filter dataset in range [-1,1]
And finally use 32000 video to train the syncnet. I also use bceloss and relu. But the loss only can drop to 0.5. Can you give me some suggestion? And how long do you use to drop loss to 0.3? @MingZJU

**Comment by MingZJU** [comment]
I used Mandarin video dataset collected by my partners. There are around 10,000 video clips with less than 1000 people. I think bceloss and relu is OK. The syncnet loss should drop soon below 0.5, after a few or decades of epoches.you may check your data again and try to modified the lr. you can also try to scale up audio encoder, though I haven't find its positive effects.-------- 原始邮件 --------发件人： Curisan ***@***.***>日期： 2023年1月30日周一 09:33收件人： primepake/wav2lip_288x288 ***@***.***>抄送： MingZJU ***@***.***>, Mention ***@***.***>主    题： Re: [primepake/wav2lip_288x288] Training with new dataset (Issue #10)
Thanks. I use AVSpeech dataset. And I did the following:
choose the 25fps videousing syncnet_python to filter dataset in range [-1,1]
And finally use 32000 video to train the syncnet. I also use bceloss and relu. But the loss only can drop to 0.5. Can you give me some suggestion? And how long do you use to drop loss to 0.3? @MingZJU

**Comment by Curisan** [comment]
Thanks, Also how many epoches do you use to drop loss below 0.3? @MingZJU

**Comment by NikitaKononov** [comment]
> Thanks, Also how many epoches do you use to drop loss below 0.3? @MingZJU

It takes 750-1250 epochs to get there, depending on you data, optimizer and LR

**Comment by Curisan** [comment]
Thanks for your suggestion. I use Relu, BCELoss, and AdamW and the train loss can drop below to 0.3. But the eval loss is about 0.42 and the eval loss value seems to be increasing when train loss decrease. It seems to be overfitting. Now I use 32000 video to train the syncnet. Did you meet the problem?  Can you give me some suggestion? @MingZJU @NikitaKononov

**Comment by humaohai** [comment]
> 2\. using syncnet_python to filter dataset in range [-1,1]

hello i was stoped at "using syncnet_python to filter dataset in range [-1,1]" ,can you paste your process code ?

**Comment by kike-0304** [comment]
> 谢谢。我使用 AVSpeech 数据集。我做了以下事情：
> 
> 1. 选择25fps视频
> 2. 使用syncnet_python过滤范围[-1,1]内的数据集
>    ，最后使用32000视频来训练syncnet。我还使用 bceloss 和 relu。但损失只能降至0.5。你能给我一些建议吗？你需要多长时间才能将损失降至 0.3？@MingZJU



> Thanks. I use AVSpeech dataset. And I did the following:
> 
> 1. choose the 25fps video
> 2. using syncnet_python to filter dataset in range [-1,1]
>    And finally use 32000 video to train the syncnet. I also use bceloss and relu. But the loss only can drop to 0.5. Can you give me some suggestion? And how long do you use to drop loss to 0.3? @MingZJU

May I ask how you use syncnet_ Python batch filters datasets within the range of [-1,1], and I have thousands of videos

**Comment by Yyyyyyxh** [comment]
> > 谢谢。我使用 AVSpeech 数据集。我做了以下事情：
> > 
> > 1. 选择25fps视频
> > 2. 使用syncnet_python过滤范围[-1,1]内的数据集
> >    ，最后使用32000视频来训练syncnet。我还使用 bceloss 和 relu。但损失只能降至0.5。你能给我一些建议吗？你需要多长时间才能将损失降至 0.3？@MingZJU
> 
> > Thanks. I use AVSpeech dataset. And I did the following:
> > 
> > 1. choose the 25fps video
> > 2. using syncnet_python to filter dataset in range [-1,1]
> >    And finally use 32000 video to train the syncnet. I also use bceloss and relu. But the loss only can drop to 0.5. Can you give me some suggestion? And how long do you use to drop loss to 0.3? @MingZJU
> 
> May I ask how you use syncnet_ Python batch filters datasets within the range of [-1,1], and I have thousands of videos

请问您知道如何筛选数据集吗，如何使用syncnet_Python

**Comment by tailangjun** [comment]
> > > 谢谢。我使用 AVSpeech 数据集。我做了以下事情：
> > > 
> > > 1. 选择25fps视频
> > > 2. 使用syncnet_python过滤范围[-1,1]内的数据集
> > >    ，最后使用32000视频来训练syncnet。我还使用 bceloss 和 relu。但损失只能降至0.5。你能给我一些建议吗？你需要多长时间才能将损失降至 0.3？@MingZJU
> > 
> > 
> > > Thanks. I use AVSpeech dataset. And I did the following:
> > > 
> > > 1. choose the 25fps video
> > > 2. using syncnet_python to filter dataset in range [-1,1]
> > >    And finally use 32000 video to train the syncnet. I also use bceloss and relu. But the loss only can drop to 0.5. Can you give me some suggestion? And how long do you use to drop loss to 0.3? @MingZJU
> > 
> > 
> > May I ask how you use syncnet_ Python batch filters datasets within the range of [-1,1], and I have thousands of videos
> 
> 请问您知道如何筛选数据集吗，如何使用syncnet_Python

我使用 bash calculate_scores_real_videos.sh you_folder_name，这样会在 syncnet_python目录下生成一个 all_scores.txt，我执行后的内容如下，都不在范围 [-1,1]内，我怀疑这个准确性有问题

9.468239, 4.9614334
9.298281, 4.6151266
9.524901, 4.234804
8.914015, 4.376094
10.106919, 2.514515
8.444702, 5.7469053
10.357641, 3.8259497
9.61246, 4.3974648
8.852642, 5.314886
9.698398, 4.9237137
9.450228, 5.185259
9.188685, 5.027153
8.574854, 4.8232346
10.000668, 5.435586
8.850996, 5.3223925
8.947002, 5.4222
9.199788, 4.866008
7.994046, 6.084631
8.054582, 6.2386484
10.574249, 2.8440237
9.321244, 4.795843
